In [1]:
import pandas as pd
import numpy as np

In [2]:
energy = pd.read_csv('energy.csv')

In [3]:
energy

Entity  Year  Access to electricity (% of population)  \
0     Afghanistan  2000                                 1.613591   
1     Afghanistan  2001                                 4.074574   
2     Afghanistan  2002                                 9.409158   
3     Afghanistan  2003                                14.738506   
4     Afghanistan  2004                                20.064968   
...           ...   ...                                      ...   
3644     Zimbabwe  2016                                42.561730   
3645     Zimbabwe  2017                                44.178635   
3646     Zimbabwe  2018                                45.572647   
3647     Zimbabwe  2019                                46.781475   
3648     Zimbabwe  2020                                52.747670   

      Access to clean fuels for cooking  \
0                                   6.2   
1                                   7.2   
2                                   8.2   
3                                   9.5   
4                                  10.9   
...                                 ...   
3644                               29.8   
3645                               29.8   
3646                               29.9   
3647                               30.1   
3648                               30.4   

      Renewable-electricity-generating-capacity-per-capita  \
0                                                  9.22      
1                                                  8.86      
2                                                  8.47      
3                                                  8.09      
4                                                  7.75      
...                                                 ...      
3644                                              62.88      
3645                                              62.33      
3646                                              82.53      
3647                                              81.40      
3648                                              80.61      

      Financial flows to developing countries (US $)  \
0                                            20000.0   
1                                           130000.0   
2                                          3950000.0   
3                                         25970000.0   
4                                                NaN   
...                                              ...   
3644                                         30000.0   
3645                                       5570000.0   
3646                                         10000.0   
3647                                        250000.0   
3648                                         30000.0   

      Renewable energy share in the total final energy consumption (%)  \
0                                                 44.99                  
1                                                 45.60                  
2                                                 37.83                  
3                                                 36.66                  
4                                                 44.24                  
...                                                 ...                  
3644                                              81.90                  
3645                                              82.46                  
3646                                              80.23                  
3647                                              81.50                  
3648                                              81.90                  

      Electricity from fossil fuels (TWh)  Electricity from nuclear (TWh)  \
0                                    0.16                             0.0   
1                                    0.09                             0.0   
2                                    0.13                             0.0   
3                                    0.31                     

In [4]:
energy['Entity'].nunique()

176

There are 176 countries in the data.

In [5]:
missing = pd.concat([energy.isnull().sum(), 100 * energy.isnull().mean()], axis=1)
missing.columns=['count', '%']
missing = missing.sort_values(by='%', ascending=False)

In [6]:
missing

count          %
Renewables (% equivalent primary energy)             2137  58.563990
Financial flows to developing countries (US $)       2089  57.248561
Renewable-electricity-generating-capacity-per-c...    931  25.513839
Value_co2_emissions_kt_by_country                     428  11.729241
gdp_growth                                            317   8.687312
gdp_per_capita                                        282   7.728145
Energy intensity level of primary energy (MJ/$2...    207   5.672787
Renewable energy share in the total final energ...    194   5.316525
Access to clean fuels for cooking                     169   4.631406
Electricity from nuclear (TWh)                        126   3.453001
Low-carbon electricity (% electricity)                 42   1.151000
Electricity from renewables (TWh)                      21   0.575500
Electricity from fossil fuels (TWh)                    21   0.575500
Access to electricity (% of population)                10   0.274048
Latitude                                                1   0.027405
Land Area(Km2)                                          1   0.027405
Density(P/Km2)                                          1   0.027405
Longitude                                               1   0.027405
Primary energy consumption per capita (kWh/person)      0   0.000000
Year                                                    0   0.000000
Entity                                                  0   0.000000

Renewables%Primary and FinancialFlows have the most missing values.  

The first thing to do is figure out how to handle the missing data.

In [7]:
missing_renewables = energy[energy['Renewables (% equivalent primary energy)'].isnull()]

In [8]:
missing_renewables['Entity'].nunique()

104

In [9]:
missing_renewables['Entity'].unique()

array(['Afghanistan', 'Albania', 'Angola', 'Antigua and Barbuda',
       'Armenia', 'Aruba', 'Bahamas', 'Bahrain', 'Barbados', 'Belize',
       'Benin', 'Bermuda', 'Bhutan', 'Bosnia and Herzegovina', 'Botswana',
       'Burkina Faso', 'Burundi', 'Cambodia', 'Cameroon',
       'Cayman Islands', 'Central African Republic', 'Chad', 'Comoros',
       'Congo', 'Costa Rica', 'Cuba', 'Djibouti', 'Dominica',
       'Dominican Republic', 'El Salvador', 'Equatorial Guinea',
       'Eritrea', 'Eswatini', 'Ethiopia', 'Fiji', 'French Guiana',
       'Gabon', 'Gambia', 'Georgia', 'Ghana', 'Grenada', 'Guatemala',
       'Guinea', 'Guinea-Bissau', 'Guyana', 'Haiti', 'Honduras',
       'Jamaica', 'Jordan', 'Kenya', 'Kiribati', 'Kyrgyzstan', 'Lebanon',
       'Lesotho', 'Liberia', 'Libya', 'Madagascar', 'Malawi', 'Maldives',
       'Mali', 'Malta', 'Mauritania', 'Mauritius', 'Mongolia',
       'Montenegro', 'Mozambique', 'Myanmar', 'Namibia', 'Nauru', 'Nepal',
       'New Caledonia', 'Nicaragua', 'Niger

In [10]:
non_missing_renewables = energy[~energy['Renewables (% equivalent primary energy)'].isnull()]

In [11]:
non_missing_renewables['Entity'].nunique()

72

In [12]:
non_missing_renewables['Entity'].unique()

array(['Algeria', 'Argentina', 'Australia', 'Austria', 'Azerbaijan',
       'Bangladesh', 'Belarus', 'Belgium', 'Brazil', 'Bulgaria', 'Canada',
       'Chile', 'China', 'Colombia', 'Croatia', 'Cyprus', 'Czechia',
       'Denmark', 'Ecuador', 'Egypt', 'Estonia', 'Finland', 'France',
       'Germany', 'Greece', 'Hungary', 'Iceland', 'India', 'Indonesia',
       'Iraq', 'Ireland', 'Israel', 'Italy', 'Japan', 'Kazakhstan',
       'Kuwait', 'Latvia', 'Lithuania', 'Luxembourg', 'Malaysia',
       'Mexico', 'Morocco', 'Netherlands', 'New Zealand',
       'North Macedonia', 'Norway', 'Oman', 'Pakistan', 'Peru',
       'Philippines', 'Poland', 'Portugal', 'Qatar', 'Romania',
       'Saudi Arabia', 'Singapore', 'Slovakia', 'Slovenia',
       'South Africa', 'Spain', 'Sri Lanka', 'Sweden', 'Switzerland',
       'Thailand', 'Trinidad and Tobago', 'Turkey', 'Turkmenistan',
       'Ukraine', 'United Arab Emirates', 'United Kingdom',
       'United States', 'Uzbekistan'], dtype=object)

For Renewables%Primary, there are 104 countries with this value missing, and 72 who have this value intact. The feature is defined as "Equivalent primary energy that is derived from renewable sources." This metric considers the proportion of renewable energy sources (like wind, solar, hydro, biomass) in the total primary energy supply. It looks at how much of the raw energy extracted from natural resources comes from renewables, before processing. This feature showcases how efficienctly a country is able to use their renewable resources, which could affect the target of GDP per capita. I don't think this data is represented through any other feature in the set. Therefore, I will not include the countries who are missing this feature in the final dataset, and only focus on the 72 who have it.


In [13]:
missing_financial_flows = energy[energy['Financial flows to developing countries (US $)'].isnull()]

In [14]:
non_missing_financial_flows = energy[~energy['Financial flows to developing countries (US $)'].isnull()]

In [15]:
missing_financial_flows['Entity'].nunique()

176

In [16]:
missing_financial_flows['Entity'].unique()

array(['Afghanistan', 'Albania', 'Algeria', 'Angola',
       'Antigua and Barbuda', 'Argentina', 'Armenia', 'Aruba',
       'Australia', 'Austria', 'Azerbaijan', 'Bahamas', 'Bahrain',
       'Bangladesh', 'Barbados', 'Belarus', 'Belgium', 'Belize', 'Benin',
       'Bermuda', 'Bhutan', 'Bosnia and Herzegovina', 'Botswana',
       'Brazil', 'Bulgaria', 'Burkina Faso', 'Burundi', 'Cambodia',
       'Cameroon', 'Canada', 'Cayman Islands', 'Central African Republic',
       'Chad', 'Chile', 'China', 'Colombia', 'Comoros', 'Congo',
       'Costa Rica', 'Croatia', 'Cuba', 'Cyprus', 'Czechia', 'Denmark',
       'Djibouti', 'Dominica', 'Dominican Republic', 'Ecuador', 'Egypt',
       'El Salvador', 'Equatorial Guinea', 'Eritrea', 'Estonia',
       'Eswatini', 'Ethiopia', 'Fiji', 'Finland', 'France',
       'French Guiana', 'Gabon', 'Gambia', 'Georgia', 'Germany', 'Ghana',
       'Greece', 'Grenada', 'Guatemala', 'Guinea', 'Guinea-Bissau',
       'Guyana', 'Haiti', 'Honduras', 'Hungary', 'Icelan

In [17]:
non_missing_financial_flows['Entity'].nunique()

118

In [18]:
non_missing_financial_flows['Entity'].unique()

array(['Afghanistan', 'Algeria', 'Angola', 'Antigua and Barbuda',
       'Argentina', 'Armenia', 'Azerbaijan', 'Bahamas', 'Bangladesh',
       'Barbados', 'Belize', 'Benin', 'Bhutan', 'Botswana', 'Brazil',
       'Burkina Faso', 'Burundi', 'Cambodia', 'Cameroon',
       'Central African Republic', 'Chad', 'Chile', 'China', 'Colombia',
       'Comoros', 'Congo', 'Costa Rica', 'Cuba', 'Djibouti', 'Dominica',
       'Dominican Republic', 'Ecuador', 'Egypt', 'El Salvador',
       'Equatorial Guinea', 'Eritrea', 'Eswatini', 'Ethiopia', 'Fiji',
       'Gabon', 'Gambia', 'Georgia', 'Ghana', 'Grenada', 'Guatemala',
       'Guinea', 'Guinea-Bissau', 'Guyana', 'Haiti', 'Honduras', 'India',
       'Indonesia', 'Iraq', 'Jamaica', 'Jordan', 'Kazakhstan', 'Kenya',
       'Kiribati', 'Kyrgyzstan', 'Lebanon', 'Lesotho', 'Liberia',
       'Madagascar', 'Malawi', 'Malaysia', 'Maldives', 'Mali',
       'Mauritania', 'Mauritius', 'Mexico', 'Mongolia', 'Morocco',
       'Mozambique', 'Myanmar', 'Namibia', 

For FinancialFlows, all countries have missing values, some more than others. I decide to drop this feature from the dataset as the irregularity for which it appears could skew analysis results.


In [19]:
energy.dropna(subset=["Renewables (% equivalent primary energy)"], inplace=True)

In [20]:
energy.drop(columns = 'Financial flows to developing countries (US $)', inplace = True)

In [21]:
energy.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1512 entries, 42 to 3564
Data columns (total 20 columns):
 #   Column                                                            Non-Null Count  Dtype  
---  ------                                                            --------------  -----  
 0   Entity                                                            1512 non-null   object 
 1   Year                                                              1512 non-null   int64  
 2   Access to electricity (% of population)                           1512 non-null   float64
 3   Access to clean fuels for cooking                                 1491 non-null   float64
 4   Renewable-electricity-generating-capacity-per-capita              693 non-null    float64
 5   Renewable energy share in the total final energy consumption (%)  1440 non-null   float64
 6   Electricity from fossil fuels (TWh)                               1512 non-null   float64
 7   Electricity from nuclear (TWh)       

The final feature with a significant amount of missing values is RenewableCapacityPerCapita, defined as "Installed Renewable energy capacity per person".

In [22]:
missing_renewable_capacity = energy[energy['Renewable-electricity-generating-capacity-per-capita'].isnull()]

In [23]:
non_missing_renewable_capacity = energy[~energy['Renewable-electricity-generating-capacity-per-capita'].isnull()]

In [24]:
missing_renewable_capacity['Entity'].nunique()

39

In [25]:
missing_renewable_capacity['Entity'].unique()

array(['Australia', 'Austria', 'Belarus', 'Belgium', 'Bulgaria', 'Canada',
       'Croatia', 'Cyprus', 'Czechia', 'Denmark', 'Estonia', 'Finland',
       'France', 'Germany', 'Greece', 'Hungary', 'Iceland', 'Ireland',
       'Israel', 'Italy', 'Japan', 'Latvia', 'Lithuania', 'Luxembourg',
       'Netherlands', 'New Zealand', 'North Macedonia', 'Norway',
       'Poland', 'Portugal', 'Romania', 'Slovakia', 'Slovenia', 'Spain',
       'Sweden', 'Switzerland', 'Ukraine', 'United Kingdom',
       'United States'], dtype=object)

In [26]:
non_missing_renewable_capacity['Entity'].nunique()

33

In [27]:
non_missing_renewable_capacity['Entity'].unique()

array(['Algeria', 'Argentina', 'Azerbaijan', 'Bangladesh', 'Brazil',
       'Chile', 'China', 'Colombia', 'Ecuador', 'Egypt', 'India',
       'Indonesia', 'Iraq', 'Kazakhstan', 'Kuwait', 'Malaysia', 'Mexico',
       'Morocco', 'Oman', 'Pakistan', 'Peru', 'Philippines', 'Qatar',
       'Saudi Arabia', 'Singapore', 'South Africa', 'Sri Lanka',
       'Thailand', 'Trinidad and Tobago', 'Turkey', 'Turkmenistan',
       'United Arab Emirates', 'Uzbekistan'], dtype=object)

There is a relatively even split between countries who have this feature and countries who do not, at 33 and 39 of the chosen 72. This feature is somewhat redundant, because total electricity from renewables already exists. For now, I will remove this column, and can always find population data of countries to create per capita features as I see fit.

In [28]:
energy.drop(columns = 'Renewable-electricity-generating-capacity-per-capita', inplace = True)

In [29]:
energy.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1512 entries, 42 to 3564
Data columns (total 19 columns):
 #   Column                                                            Non-Null Count  Dtype  
---  ------                                                            --------------  -----  
 0   Entity                                                            1512 non-null   object 
 1   Year                                                              1512 non-null   int64  
 2   Access to electricity (% of population)                           1512 non-null   float64
 3   Access to clean fuels for cooking                                 1491 non-null   float64
 4   Renewable energy share in the total final energy consumption (%)  1440 non-null   float64
 5   Electricity from fossil fuels (TWh)                               1512 non-null   float64
 6   Electricity from nuclear (TWh)                                    1407 non-null   float64
 7   Electricity from renewables (TWh)    

After dealing with major missing data issues, the difference between the fully intact columns vs imperfect columns is 132 rows (1512 - 1380). This difference is not significant enough to where dropping all nan rows would cripple the dataset. Still, I will keep that as a last resort and look into various forms of data imputation. Imputation can take on many forms, and will vary on a feature to feature basis.

In [30]:
missing = pd.concat([energy.isnull().sum(), 100 * energy.isnull().mean()], axis=1)
missing.columns=['count', '%']
missing = missing.sort_values(by='%', ascending=False)

In [31]:
missing

count         %
Value_co2_emissions_kt_by_country                     132  8.730159
Electricity from nuclear (TWh)                        105  6.944444
Renewable energy share in the total final energ...     72  4.761905
Energy intensity level of primary energy (MJ/$2...     72  4.761905
gdp_growth                                             65  4.298942
gdp_per_capita                                         64  4.232804
Access to clean fuels for cooking                      21  1.388889
Entity                                                  0  0.000000
Renewables (% equivalent primary energy)                0  0.000000
Latitude                                                0  0.000000
Land Area(Km2)                                          0  0.000000
Density(P/Km2)                                          0  0.000000
Primary energy consumption per capita (kWh/person)      0  0.000000
Year                                                    0  0.000000
Low-carbon electricity (% electricity)                  0  0.000000
Electricity from renewables (TWh)                       0  0.000000
Electricity from fossil fuels (TWh)                     0  0.000000
Access to electricity (% of population)                 0  0.000000
Longitude                                               0  0.000000

value_co2_emissions has 132 missing values.

In [32]:
missing_counts = energy.groupby('Entity')['Value_co2_emissions_kt_by_country'].apply(lambda x: x.isnull().sum()).sort_values(ascending = False)
sorted_entities = missing_counts.reset_index(name='Missing_Values_Count')
print(sorted_entities)

        Entity  Missing_Values_Count
0     Slovakia                    21
1       Turkey                    21
2        Egypt                    21
3      Algeria                     1
4     Pakistan                     1
..         ...                   ...
67      Greece                     1
68     Germany                     1
69      France                     1
70     Finland                     1
71  Uzbekistan                     1

[72 rows x 2 columns]


Egypt, Turkey, and Slovakia are all missing this feature for the entire timeframe. A quick google search links to https://data.worldbank.org/indicator/EN.ATM.CO2E.KT?end=2020&locations=SK&name_desc=false&start=1990&view=chart, which is a primary source for the dataset.

It seems the data is available for these countries, but under different names than what the compiler probably expected to see:

1. Egypt, Arab Republic
2. Turkiye
3. Slovak Republic

These are not the typical names associated with these countries in the US (to my experience), showing how information can be lost due to semantics. I manually add in the data for these three countries.

In [33]:
co2emissions = pd.read_csv("co2emissions.csv")
countries_to_filter = ['Turkiye', 'Egypt, Arab Rep.', 'Slovak Republic']
co2_add_ons = co2emissions[co2emissions['Country Name'].isin(countries_to_filter)]
columns_to_keep = ['Country Name'] + [str(year) for year in range(2000, 2021)]
filtered_df = co2_add_ons.filter(columns_to_keep)
melted_df = filtered_df.melt(id_vars='Country Name', var_name='Year', value_name='CO2 Emission')
sorted_co2 = melted_df.sort_values(by=['Country Name', 'Year'])
sorted_co2['Country Name'] = sorted_co2['Country Name'].replace('Egypt, Arab Rep.', 'Egypt')
sorted_co2['Country Name'] = sorted_co2['Country Name'].replace('Slovak Republic', 'Slovakia')
sorted_co2['Country Name'] = sorted_co2['Country Name'].replace('Turkiye', 'Turkey')
sorted_co2 = sorted_co2.rename(columns={'Country Name': 'Entity'})
sorted_co2 = sorted_co2.rename(columns={'CO2 Emission': 'Value_co2_emissions_kt_by_country'})
sorted_co2['Year'] = sorted_co2['Year'].astype(int)

In [34]:
sorted_co2

Entity  Year  Value_co2_emissions_kt_by_country
0    Egypt  2000                           114614.4
3    Egypt  2001                           126704.4
6    Egypt  2002                           129441.2
9    Egypt  2003                           133020.4
12   Egypt  2004                           144503.9
..     ...   ...                                ...
50  Turkey  2016                           376399.0
53  Turkey  2017                           418098.2
56  Turkey  2018                           414111.9
59  Turkey  2019                           398772.9
62  Turkey  2020                           407406.2

[63 rows x 3 columns]

In [35]:
energy.set_index(['Entity', 'Year'], inplace=True)
sorted_co2.set_index(['Entity', 'Year'], inplace=True)
energy.update(sorted_co2)
energy.reset_index(inplace=True)

In [36]:
missing_counts = energy.groupby('Entity')['Value_co2_emissions_kt_by_country'].apply(lambda x: x.isnull().sum()).sort_values(ascending = False)
sorted_entities = missing_counts.reset_index(name='Missing_Values_Count')
print(sorted_entities)

             Entity  Missing_Values_Count
0           Algeria                     1
1   North Macedonia                     1
2            Poland                     1
3       Philippines                     1
4              Peru                     1
..              ...                   ...
67          Estonia                     1
68       Uzbekistan                     1
69            Egypt                     0
70           Turkey                     0
71         Slovakia                     0

[72 rows x 2 columns]


The 63 missing rows of co2 emissions for the three countries have been taken care of. For the rest of missing emissions, I will impute with the same value from the previous year since countries arte only missing a single year.

In [37]:
energy['Value_co2_emissions_kt_by_country'] = energy.groupby('Entity')['Value_co2_emissions_kt_by_country'].fillna(method='ffill')

In [38]:
missing_counts = energy.groupby('Entity')['Value_co2_emissions_kt_by_country'].apply(lambda x: x.isnull().sum()).sort_values(ascending = False)
sorted_entities = missing_counts.reset_index(name='Missing_Values_Count')
print(sorted_entities)

        Entity  Missing_Values_Count
0      Algeria                     0
1    Argentina                     0
2        Qatar                     0
3     Portugal                     0
4       Poland                     0
..         ...                   ...
67     Germany                     0
68      France                     0
69     Finland                     0
70     Estonia                     0
71  Uzbekistan                     0

[72 rows x 2 columns]


All taken care of.

Next is electricity from nuclear, with 105 missing values.

In [39]:
missing_counts = energy.groupby('Entity')['Electricity from nuclear (TWh)'].apply(lambda x: x.isnull().sum()).sort_values(ascending = False)
sorted_entities = missing_counts.reset_index(name='Missing_Values_Count')
print(sorted_entities)

          Entity  Missing_Values_Count
0       Malaysia                    21
1   Saudi Arabia                    21
2          Chile                    21
3      Indonesia                    21
4     Kazakhstan                    21
..           ...                   ...
67        Israel                     0
68         Italy                     0
69         Japan                     0
70        Kuwait                     0
71    Uzbekistan                     0

[72 rows x 2 columns]


5 countries are missing values for this feature in its entirety.

In [40]:
entities_with_zero_nuclear = energy[energy['Electricity from nuclear (TWh)'] == 0]['Entity']
print("Entities with Electricity from nuclear (TWh) value of 0:")
print(entities_with_zero_nuclear.unique())

Entities with Electricity from nuclear (TWh) value of 0:
['Algeria' 'Australia' 'Austria' 'Azerbaijan' 'Bangladesh' 'Belarus'
 'Colombia' 'Croatia' 'Cyprus' 'Denmark' 'Ecuador' 'Egypt' 'Estonia'
 'Greece' 'Iceland' 'Iraq' 'Ireland' 'Israel' 'Italy' 'Japan' 'Kuwait'
 'Latvia' 'Lithuania' 'Luxembourg' 'Morocco' 'New Zealand'
 'North Macedonia' 'Norway' 'Oman' 'Peru' 'Philippines' 'Poland'
 'Portugal' 'Qatar' 'Singapore' 'Sri Lanka' 'Thailand'
 'Trinidad and Tobago' 'Turkey' 'Turkmenistan' 'United Arab Emirates'
 'Uzbekistan']


It seems these 5 countries (Malaysia, Saudi Arabia, Chile, Indonesia, Kazakhstan) are true nan values, because there are several other countries with values of 0 who are documented properly, and I don't see their names here.

I looked into the nuclear energy programs for the 5, and found the following:

1. Malaysia - "In the past, the country had an ambitious nuclear program with plans to establish two operational nuclear power plants by 2021. However, these plans took an unexpected turn and were indefinitely postponed, ultimately leading to the disbandment of the Malaysia Nuclear Power Cooperation in 2018." (https://www.nuclearbusiness-platform.com/media/insights/nuclear-potentially-can-help-malaysia-to-become-clean-energy-hub#:~:text=In%20the%20past%2C%20the%20country,Nuclear%20Power%20Cooperation%20in%202018.)

2. Saudi Arabia - "Saudi Arabia has plans to establish a civil nuclear power industry. The country in 2013 projected 17 GWe of nuclear capacity by 2032, but that target was abandoned two years later. At present, virtually all of the country's electricity is generated by burning fossil fuels." (https://world-nuclear.org/information-library/country-profiles/countries-o-s/saudi-arabia)

3. Chile - "Currently, Chile has no nuclear power plants and has not taken any decision regarding the development of a nuclear power programme. However, Chile has a strong background on other nuclear applications." (https://www-pub.iaea.org/MTCD/Publications/PDF/cnpp2018/countryprofiles/Chile/Chile.htm#:~:text=Currently%2C%20Chile%20has%20no%20nuclear,background%20on%20other%20nuclear%20applications.)

4. Indonesia - as of March 2023, "The USA and Indonesia have announced a strategic partnership to help Indonesia develop its nuclear energy programme, supporting Indonesia's interest in deploying small modular reactor (SMR) technology to meet its energy security and climate goals." (https://world-nuclear-news.org/Articles/USA,-Indonesia-announce-partnership-on-SMRs)

5. Kazakhstan - as of September 2023, "Kazakhstan is to hold a referendum on the construction of a nuclear power plant in the country, President Kassym-Jomart Tokayev announced during a state-of-the-nation address. The date of the vote has yet to be determined." (https://world-nuclear-news.org/Articles/Kazakh-people-to-decide-on-nuclear-plant-construct)

These countries do not currently have operable nuclear power in their countries, but some are looking to investigate the possibility.  

Imputing the missing data with 0 makes the most sense for these 5 countries.




In [41]:
energy['Electricity from nuclear (TWh)'].fillna(0, inplace=True)

In [42]:
missing = pd.concat([energy.isnull().sum(), 100 * energy.isnull().mean()], axis=1)
missing.columns=['count', '%']
missing = missing.sort_values(by='%', ascending=False)
missing

count         %
Renewable energy share in the total final energ...     72  4.761905
Energy intensity level of primary energy (MJ/$2...     72  4.761905
gdp_growth                                             65  4.298942
gdp_per_capita                                         64  4.232804
Access to clean fuels for cooking                      21  1.388889
Entity                                                  0  0.000000
Value_co2_emissions_kt_by_country                       0  0.000000
Latitude                                                0  0.000000
Land Area(Km2)                                          0  0.000000
Density(P/Km2)                                          0  0.000000
Renewables (% equivalent primary energy)                0  0.000000
Primary energy consumption per capita (kWh/person)      0  0.000000
Year                                                    0  0.000000
Low-carbon electricity (% electricity)                  0  0.000000
Electricity from renewables (TWh)                       0  0.000000
Electricity from nuclear (TWh)                          0  0.000000
Electricity from fossil fuels (TWh)                     0  0.000000
Access to electricity (% of population)                 0  0.000000
Longitude                                               0  0.000000

In [43]:
missing_counts = energy.groupby('Entity')['Renewable energy share in the total final energy consumption (%)'].apply(lambda x: x.isnull().sum()).sort_values(ascending = False)
sorted_entities = missing_counts.reset_index(name='Missing_Values_Count')
print(sorted_entities)

        Entity  Missing_Values_Count
0      Algeria                     1
1    Argentina                     1
2        Qatar                     1
3     Portugal                     1
4       Poland                     1
..         ...                   ...
67     Germany                     1
68      France                     1
69     Finland                     1
70     Estonia                     1
71  Uzbekistan                     1

[72 rows x 2 columns]


In [44]:
missing_counts = energy.groupby('Entity')['Energy intensity level of primary energy (MJ/$2017 PPP GDP)'].apply(lambda x: x.isnull().sum()).sort_values(ascending = False)
sorted_entities = missing_counts.reset_index(name='Missing_Values_Count')
print(sorted_entities)

        Entity  Missing_Values_Count
0      Algeria                     1
1    Argentina                     1
2        Qatar                     1
3     Portugal                     1
4       Poland                     1
..         ...                   ...
67     Germany                     1
68      France                     1
69     Finland                     1
70     Estonia                     1
71  Uzbekistan                     1

[72 rows x 2 columns]


For these two features, it looks like just a single year is missing for every country, so I will impute with the previous value.

In [45]:
energy['Renewable energy share in the total final energy consumption (%)'] = energy.groupby('Entity')['Renewable energy share in the total final energy consumption (%)'].fillna(method='ffill')
energy['Energy intensity level of primary energy (MJ/$2017 PPP GDP)'] = energy.groupby('Entity')['Energy intensity level of primary energy (MJ/$2017 PPP GDP)'].fillna(method='ffill')

In [46]:
missing = pd.concat([energy.isnull().sum(), 100 * energy.isnull().mean()], axis=1)
missing.columns=['count', '%']
missing = missing.sort_values(by='%', ascending=False)
missing

count         %
gdp_growth                                             65  4.298942
gdp_per_capita                                         64  4.232804
Access to clean fuels for cooking                      21  1.388889
Entity                                                  0  0.000000
Energy intensity level of primary energy (MJ/$2...      0  0.000000
Latitude                                                0  0.000000
Land Area(Km2)                                          0  0.000000
Density(P/Km2)                                          0  0.000000
Renewables (% equivalent primary energy)                0  0.000000
Value_co2_emissions_kt_by_country                       0  0.000000
Primary energy consumption per capita (kWh/person)      0  0.000000
Year                                                    0  0.000000
Low-carbon electricity (% electricity)                  0  0.000000
Electricity from renewables (TWh)                       0  0.000000
Electricity from nuclear (TWh)                          0  0.000000
Electricity from fossil fuels (TWh)                     0  0.000000
Renewable energy share in the total final energ...      0  0.000000
Access to electricity (% of population)                 0  0.000000
Longitude                                               0  0.000000

Just three more imperfect features to go.

In [47]:
missing_counts = energy.groupby('Entity')['gdp_growth'].apply(lambda x: x.isnull().sum()).sort_values(ascending = False)
sorted_entities = missing_counts.reset_index(name='Missing_Values_Count')
print(sorted_entities)

          Entity  Missing_Values_Count
0          Egypt                    21
1       Slovakia                    21
2        Czechia                    21
3          Qatar                     1
4   Turkmenistan                     1
..           ...                   ...
67        Israel                     0
68         Italy                     0
69         Japan                     0
70    Kazakhstan                     0
71    Uzbekistan                     0

[72 rows x 2 columns]


In [48]:
missing_counts = energy.groupby('Entity')['gdp_per_capita'].apply(lambda x: x.isnull().sum()).sort_values(ascending = False)
sorted_entities = missing_counts.reset_index(name='Missing_Values_Count')
print(sorted_entities)

          Entity  Missing_Values_Count
0          Egypt                    21
1       Slovakia                    21
2        Czechia                    21
3   Turkmenistan                     1
4       Pakistan                     0
..           ...                   ...
67        Israel                     0
68         Italy                     0
69         Japan                     0
70    Kazakhstan                     0
71    Uzbekistan                     0

[72 rows x 2 columns]


Once again, Egypt and Slovakia are missing data, and now Czehcia as well. I will find this information from the World Bank and update the main dataframe.

In [49]:
gdpgrowth = pd.read_csv("gdpgrowth.csv")
countries_to_filter = ['Czechia', 'Egypt, Arab Rep.', 'Slovak Republic']
gdpgrowth_add_ons = gdpgrowth[gdpgrowth['Country Name'].isin(countries_to_filter)]
columns_to_keep = ['Country Name'] + [str(year) for year in range(2000, 2021)]
filtered_df = gdpgrowth_add_ons.filter(columns_to_keep)
melted_df = filtered_df.melt(id_vars='Country Name', var_name='Year', value_name='gdp_growth')
sorted_gdpgrowth = melted_df.sort_values(by=['Country Name', 'Year'])
sorted_gdpgrowth['Country Name'] = sorted_gdpgrowth['Country Name'].replace('Egypt, Arab Rep.', 'Egypt')
sorted_gdpgrowth['Country Name'] = sorted_gdpgrowth['Country Name'].replace('Slovak Republic', 'Slovakia')
sorted_gdpgrowth = sorted_gdpgrowth.rename(columns={'Country Name': 'Entity'})
sorted_gdpgrowth['Year'] = sorted_gdpgrowth['Year'].astype(int)

In [50]:
energy.set_index(['Entity', 'Year'], inplace=True)
sorted_gdpgrowth.set_index(['Entity', 'Year'], inplace=True)
energy.update(sorted_gdpgrowth)
energy.reset_index(inplace=True)

In [51]:
gdpcapita = pd.read_csv("gdpcapita.csv")
countries_to_filter = ['Czechia', 'Egypt, Arab Rep.', 'Slovak Republic']
gdpcapita_add_ons = gdpcapita[gdpgrowth['Country Name'].isin(countries_to_filter)]
columns_to_keep = ['Country Name'] + [str(year) for year in range(2000, 2021)]
filtered_df = gdpcapita_add_ons.filter(columns_to_keep)
melted_df = filtered_df.melt(id_vars='Country Name', var_name='Year', value_name='gdp_per_capita')
sorted_gdpcapita = melted_df.sort_values(by=['Country Name', 'Year'])
sorted_gdpcapita['Country Name'] = sorted_gdpcapita['Country Name'].replace('Egypt, Arab Rep.', 'Egypt')
sorted_gdpcapita['Country Name'] = sorted_gdpcapita['Country Name'].replace('Slovak Republic', 'Slovakia')
sorted_gdpcapita = sorted_gdpcapita.rename(columns={'Country Name': 'Entity'})
sorted_gdpcapita['Year'] = sorted_gdpcapita['Year'].astype(int)

In [52]:
energy.set_index(['Entity', 'Year'], inplace=True)
sorted_gdpcapita.set_index(['Entity', 'Year'], inplace=True)
energy.update(sorted_gdpcapita)
energy.reset_index(inplace=True)

In [53]:
missing_counts = energy.groupby('Entity')['gdp_growth'].apply(lambda x: x.isnull().sum()).sort_values(ascending = False)
sorted_entities = missing_counts.reset_index(name='Missing_Values_Count')
print(sorted_entities)

          Entity  Missing_Values_Count
0          Qatar                     1
1   Turkmenistan                     1
2         Norway                     0
3       Portugal                     0
4         Poland                     0
..           ...                   ...
67       Germany                     0
68        France                     0
69       Finland                     0
70       Estonia                     0
71    Uzbekistan                     0

[72 rows x 2 columns]


In [54]:
missing_counts = energy.groupby('Entity')['gdp_per_capita'].apply(lambda x: x.isnull().sum()).sort_values(ascending = False)
sorted_entities = missing_counts.reset_index(name='Missing_Values_Count')
print(sorted_entities)

          Entity  Missing_Values_Count
0   Turkmenistan                     1
1        Algeria                     0
2         Norway                     0
3       Portugal                     0
4         Poland                     0
..           ...                   ...
67       Germany                     0
68        France                     0
69       Finland                     0
70       Estonia                     0
71    Uzbekistan                     0

[72 rows x 2 columns]


Much better, I'll impute with previous values.

In [55]:
energy['gdp_growth'] = energy.groupby('Entity')['gdp_growth'].fillna(method='ffill')
energy['gdp_per_capita'] = energy.groupby('Entity')['gdp_per_capita'].fillna(method='ffill')

In [56]:
missing_counts = energy.groupby('Entity')['gdp_growth'].apply(lambda x: x.isnull().sum()).sort_values(ascending = False)
sorted_entities = missing_counts.reset_index(name='Missing_Values_Count')
print(sorted_entities)

        Entity  Missing_Values_Count
0        Qatar                     1
1      Algeria                     0
2    Argentina                     0
3     Portugal                     0
4       Poland                     0
..         ...                   ...
67     Germany                     0
68      France                     0
69     Finland                     0
70     Estonia                     0
71  Uzbekistan                     0

[72 rows x 2 columns]


In [57]:
missing_counts = energy.groupby('Entity')['gdp_per_capita'].apply(lambda x: x.isnull().sum()).sort_values(ascending = False)
sorted_entities = missing_counts.reset_index(name='Missing_Values_Count')
print(sorted_entities)

        Entity  Missing_Values_Count
0      Algeria                     0
1    Argentina                     0
2        Qatar                     0
3     Portugal                     0
4       Poland                     0
..         ...                   ...
67     Germany                     0
68      France                     0
69     Finland                     0
70     Estonia                     0
71  Uzbekistan                     0

[72 rows x 2 columns]


Strange, Qatar is still missing a single value for gdp_growth.

In [58]:
qatar = energy['Entity'] == 'Qatar'

In [59]:
energy[qatar]['gdp_growth']

1092          NaN
1093     3.898187
1094     7.182152
1095     3.719959
1096    19.218915
1097     7.492758
1098    26.170246
1099    17.985657
1100    17.663556
1101    11.956561
1102    19.592332
1103    13.375176
1104     4.730012
1105     5.556041
1106     5.334323
1107     4.753346
1108     3.064192
1109    -1.497605
1110     1.234872
1111     0.688241
1112    -3.557604
Name: gdp_growth, dtype: float64

That's why, I can't impute from previous because there is none, so I'll impute with a forward fill.

In [60]:
energy['gdp_growth'] = energy.groupby('Entity')['gdp_growth'].fillna(method='bfill')

In [61]:
missing = pd.concat([energy.isnull().sum(), 100 * energy.isnull().mean()], axis=1)
missing.columns=['count', '%']
missing = missing.sort_values(by='%', ascending=False)
missing

count         %
Access to clean fuels for cooking                      21  1.388889
Entity                                                  0  0.000000
Energy intensity level of primary energy (MJ/$2...      0  0.000000
Latitude                                                0  0.000000
Land Area(Km2)                                          0  0.000000
Density(P/Km2)                                          0  0.000000
gdp_per_capita                                          0  0.000000
gdp_growth                                              0  0.000000
Renewables (% equivalent primary energy)                0  0.000000
Value_co2_emissions_kt_by_country                       0  0.000000
Primary energy consumption per capita (kWh/person)      0  0.000000
Year                                                    0  0.000000
Low-carbon electricity (% electricity)                  0  0.000000
Electricity from renewables (TWh)                       0  0.000000
Electricity from nuclear (TWh)                          0  0.000000
Electricity from fossil fuels (TWh)                     0  0.000000
Renewable energy share in the total final energ...      0  0.000000
Access to electricity (% of population)                 0  0.000000
Longitude                                               0  0.000000

In [62]:
missing_counts = energy.groupby('Entity')['Access to clean fuels for cooking'].apply(lambda x: x.isnull().sum()).sort_values(ascending = False)
sorted_entities = missing_counts.reset_index(name='Missing_Values_Count')
print(sorted_entities)

          Entity  Missing_Values_Count
0       Bulgaria                    21
1   Saudi Arabia                     0
2          Qatar                     0
3       Portugal                     0
4         Poland                     0
..           ...                   ...
67        Greece                     0
68       Germany                     0
69        France                     0
70       Finland                     0
71    Uzbekistan                     0

[72 rows x 2 columns]


For Bulgaria, the World Bank is missing this metric for the country, so I will omit the country as a whole.

In [63]:
energy = energy[energy['Entity'] != 'Bulgaria']

In [64]:
energy.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1491 entries, 0 to 1511
Data columns (total 19 columns):
 #   Column                                                            Non-Null Count  Dtype  
---  ------                                                            --------------  -----  
 0   Entity                                                            1491 non-null   object 
 1   Year                                                              1491 non-null   int64  
 2   Access to electricity (% of population)                           1491 non-null   float64
 3   Access to clean fuels for cooking                                 1491 non-null   float64
 4   Renewable energy share in the total final energy consumption (%)  1491 non-null   float64
 5   Electricity from fossil fuels (TWh)                               1491 non-null   float64
 6   Electricity from nuclear (TWh)                                    1491 non-null   float64
 7   Electricity from renewables (TWh)     

I need to convert some features to the correct dtype. I want lat, lon to be object, and for Density to be numerical.

In [65]:
energy['Latitude'] = energy['Latitude'].astype(str)
energy['Longitude'] = energy['Longitude'].astype(str)

<ipython-input-65-e532aa3f5632>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  energy['Latitude'] = energy['Latitude'].astype(str)
<ipython-input-65-e532aa3f5632>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  energy['Longitude'] = energy['Longitude'].astype(str)


In [66]:
energy['Density(P/Km2)'] = energy['Density(P/Km2)'].str.replace(',', '').astype(int)

<ipython-input-66-d778fb118386>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  energy['Density(P/Km2)'] = energy['Density(P/Km2)'].str.replace(',', '').astype(int)


In [67]:
energy.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1491 entries, 0 to 1511
Data columns (total 19 columns):
 #   Column                                                            Non-Null Count  Dtype  
---  ------                                                            --------------  -----  
 0   Entity                                                            1491 non-null   object 
 1   Year                                                              1491 non-null   int64  
 2   Access to electricity (% of population)                           1491 non-null   float64
 3   Access to clean fuels for cooking                                 1491 non-null   float64
 4   Renewable energy share in the total final energy consumption (%)  1491 non-null   float64
 5   Electricity from fossil fuels (TWh)                               1491 non-null   float64
 6   Electricity from nuclear (TWh)                                    1491 non-null   float64
 7   Electricity from renewables (TWh)     

I now have a fully intact dataframe for 71 countries. To finish the data wrangling process, I'll need to make sure all features are scaled and normalized. I'll rename some of the columns to make them more concise as well.

In [68]:
energy

Entity  Year  Access to electricity (% of population)  \
0        Algeria  2000                                 98.97310   
1        Algeria  2001                                 98.96687   
2        Algeria  2002                                 98.95306   
3        Algeria  2003                                 98.93401   
4        Algeria  2004                                 98.91208   
...          ...   ...                                      ...   
1507  Uzbekistan  2016                                100.00000   
1508  Uzbekistan  2017                                100.00000   
1509  Uzbekistan  2018                                100.00000   
1510  Uzbekistan  2019                                100.00000   
1511  Uzbekistan  2020                                100.00000   

      Access to clean fuels for cooking  \
0                                 97.10   
1                                 97.30   
2                                 97.80   
3                                 98.00   
4                                 98.20   
...                                 ...   
1507                              85.20   
1508                              84.90   
1509                              84.30   
1510                              84.60   
1511                              84.25   

      Renewable energy share in the total final energy consumption (%)  \
0                                                  0.43                  
1                                                  0.43                  
2                                                  0.51                  
3                                                  0.47                  
4                                                  0.44                  
...                                                 ...                  
1507                                               1.61                  
1508                                               1.75                  
1509                                               1.49                  
1510                                               1.57                  
1511                                               1.57                  

      Electricity from fossil fuels (TWh)  Electricity from nuclear (TWh)  \
0                                   23.84                             0.0   
1                                   24.96                             0.0   
2                                   25.94                             0.0   
3                                   27.54                             0.0   
4                                   29.14                             0.0   
...                                   ...                             ...   
1507                                48.75                             0.0   
1508                                49.71                             0.0   
1509                                53.58                             0.0   
1510                                53.64                             0.0   
1511                                50.93                             0.0   

      Electricity from renewables (TWh)  \
0                                  0.05   
1                                  0.07   
2                                  0.06   
3                                  0.26   
4                                  0.25   
...                                 ...   
1507                               7.25   
1508                               8.35   
1509                               5.85   
1510                               6.47   
1511                               5.01   

      Low-carbon electricity (% electricity)  \
0                                   0.209293   
1                                   0.279664   
2                                   0.230769   
3                                   0.935252   
4                                   0.850630   
...                                      ...   
1507                 

In [69]:
energy = energy.rename(columns={'Access to electricity (% of population)': 'electricity_access_%'})
energy = energy.rename(columns={'Access to clean fuels for cooking': 'clean_fuels_access_%'})
energy = energy.rename(columns={'Renewable energy share in the total final energy consumption (%)': 'renewable_energy_share_final_%'})
energy = energy.rename(columns={'Electricity from fossil fuels (TWh)': 'electricity_fossil_fuels_Twh'})
energy = energy.rename(columns={'Electricity from nuclear (TWh)': 'electricity_nuclear_Twh'})
energy = energy.rename(columns={'Electricity from renewables (TWh)': 'electricity_renewables_Twh'})
energy = energy.rename(columns={'Low-carbon electricity (% electricity)': 'low_carbon_electricity_%'})
energy = energy.rename(columns={'Primary energy consumption per capita (kWh/person)': 'primary_energy_consumption_per_capita_kwh'})
conversion_factor = 1 / 10**9
energy['primary_energy_consumption_per_capita_twh'] = energy['primary_energy_consumption_per_capita_kwh'] * conversion_factor
energy.drop(columns=['primary_energy_consumption_per_capita_kwh'], inplace=True)
energy = energy.rename(columns={'Energy intensity level of primary energy (MJ/$2017 PPP GDP)': 'energy_intensity_mj_ppp'})
energy = energy.rename(columns={'Value_co2_emissions_kt_by_country': 'co2_emissions_kt'})
energy = energy.rename(columns={'Renewables (% equivalent primary energy)': 'renewable_energy_share_primary_%'})

In [71]:
energy

Entity  Year  electricity_access_%  clean_fuels_access_%  \
0        Algeria  2000              98.97310                 97.10   
1        Algeria  2001              98.96687                 97.30   
2        Algeria  2002              98.95306                 97.80   
3        Algeria  2003              98.93401                 98.00   
4        Algeria  2004              98.91208                 98.20   
...          ...   ...                   ...                   ...   
1507  Uzbekistan  2016             100.00000                 85.20   
1508  Uzbekistan  2017             100.00000                 84.90   
1509  Uzbekistan  2018             100.00000                 84.30   
1510  Uzbekistan  2019             100.00000                 84.60   
1511  Uzbekistan  2020             100.00000                 84.25   

      renewable_energy_share_final_%  electricity_fossil_fuels_Twh  \
0                               0.43                         23.84   
1                               0.43                         24.96   
2                               0.51                         25.94   
3                               0.47                         27.54   
4                               0.44                         29.14   
...                              ...                           ...   
1507                            1.61                         48.75   
1508                            1.75                         49.71   
1509                            1.49                         53.58   
1510                            1.57                         53.64   
1511                            1.57                         50.93   

      electricity_nuclear_Twh  electricity_renewables_Twh  \
0                         0.0                        0.05   
1                         0.0                        0.07   
2                         0.0                        0.06   
3                         0.0                        0.26   
4                         0.0                        0.25   
...                       ...                         ...   
1507                      0.0                        7.25   
1508                      0.0                        8.35   
1509                      0.0                        5.85   
1510                      0.0                        6.47   
1511                      0.0                        5.01   

      low_carbon_electricity_%  energy_intensity_mj_ppp  co2_emissions_kt  \
0                     0.209293                     4.18       80050.00000   
1                     0.279664                     4.07       78650.00000   
2                     0.230769                     4.12       82400.00153   
3                     0.935252                     4.08       88190.00244   
4                     0.850630                     3.96       89489.99786   
...                        ...                      ...               ...   
1507                 12.946428                     7.82      105230.00340   
1508                 14.381675                     7.88      109529.99880   
1509                  9.843514                     9.05      112470.00120   
1510                 10.763600                     8.37      116709.99910   
1511                  8.956025                     8.37      116709.99910   

      renewable_energy_share_primary_%  gdp_growth  gdp_per_capita  \
0                             0.053235    3.800000     1765.027146   
1                             0.065218    3.000000     1740.606654   
2                             0.051677    5.600000     1781.828908   
3                             0.228104    7.200000     2103.381291   
4                             0.206787    4.300000     2610.185422   
...                                ...         ...             ...   
1507                          3.567936    5.932151     2704.677188   
1508                          3.972285    4.395275     1916.764642   
1509                          2.927033  

In [72]:
energy.to_csv('energy_cleaned.csv', index=False)

I now have a fully cleaned dataframe for 71 countries ready for further processing. The next step is to dive into the data and find which features are important and why, in regards to the target of gdp_per_capita.